### ISSUE TITLE
Per-request EncoderFilters

### ISSUE THREAD

**Issue by andrus** [agrestio/agrest#420]
The use case I currently have to deal with is Encoders that perform localization/translation of values depending on the language code passed in the URL. 

There are a few ways to alter how response encoding works - custom Encoders and EncoderFilters. Encoders can be redefined per request (`SelectBuilder.dataEncoder(..)` ), but filters - only once for the entire runtime.

Defining a full Encoder is useless in most situations, as they are hard to create from scratch. EncoderFilters on the other hand allow to reuse the default complex `DataResponseEncoder` object, and just intercept (change/suppress) individual property values.

So it would be great to have the ability to specify an EncoderFilter or a similar object per request. This will allow to initialize encoders with request-specific context information (or even install it optionally based on a data-driven condition), resulting in less convoluted and more performant encoders, and avoiding passing request state through ThreadLocals, etc.

**Comment by andrus** [comment]
A bit more analysis... As is often the case in Agrest, we are faced with the challenge of each request having a `ResourceEntity` tree of an arbitrary depth and structure, containing anywhere from a single root entity, to the entire app model. This was the main reason why `EncoderFilter` was previously only allowed to be defined globally. I.e. specifying rules once per entity is easier that specifying all possible per-entity filters in a request method. 

The only downside was the inability to pass the request state to encoders. 

So now we need to decide how this API should work to provide the best of both worlds (dealing with every single entity inside one request method, vs initializing filters with request state).

**Comment by andrus** [comment]
Another challenge - `EncoderFilter` is an entity-level object, so it doesn't provide an easy way to alter encoding of individual properties (as required in the original use case). So we will need to build a mechanism to reuse Agrest provided `EntityProperty` collection within `EncoderFilter`

**Comment by andrus** [comment]
## Upgrade

* `EncoderFilter` was renamed to `EntityEncoderFilter`, as the new name points to the exact place where filtering occurs (per entity object).
*  Previous `EntityEncoderFilter` that was used as an abstract superclass for the filters was removed. If you subclassed it, use `EntityEncoderFilter` static builder methods instead.

**Comment by andrus** [comment]
The issue of individual property access mentioned above is still unresolved. Will probably open a separate task.
